### PR TITLE
Search feature only

### PR DESCRIPTION
**Pull request by bscrow**
This PR contains the same set of changes in #43, minus the writeup files

### CODE REVIEW

File: pysradb/cli.py [hunk 1]
```diff
@@ -604,32 +679,129 @@ def parse_args(args=None):
...
+        default="sra",
+        help="Select the db API (sra, ena, or geo) to query, default = sra",
     )
     subparser.add_argument(
-        "--expand", action="store_true", help="Should sample_attribute be expanded"
+        "-v",
+        "--verbosity",
+        choices=[0, 1, 2, 3],
+        default=2,
+        help="Level of search result details (0, 1, 2 or 3), default = 2",
```
Comment (saketkc) [inline pysradb/cli.py:723]: Can you elaborate o what each of the verbosity level reflect in short? For example 0: experiment and run accessions only 2: records in 1 + pmids 3: ...

File: pysradb/search.py [hunk 2]
```diff
@@ -0,0 +1,1619 @@
...
+        date_regex = "(0[1-9]|[12][0-9]|3[01])-(0[1-9]|1[012])-(19|20)[0-9]{2}"
+        if self.fields["publication_date"] and not re.match(
+            f"^{date_regex}(:{date_regex})?$", self.fields["publication_date"]
+        ):
+            message += (
+                f"Incorrect publication date format: {self.fields['publication_date']}\n"
+                f"Expected --publication-date format: dd-mm-yyyy or dd-mm-yyyy:dd-mm-yyyy, between 1900-2099\n\n"
+            )
+        # verify platform
+        platform_matcher = {
```
Comment (saketkc) [inline pysradb/search.py:250]: Can you leave a comment here how these were determined? If a new platform is introduced in SRA/ENA how should this be changed?

File: pysradb/search.py [hunk 3]
```diff
@@ -0,0 +1,1619 @@
...
+        if axes == ("Base Count",):
+            count = list(self.stats["count_data"])
+            title = "Histogram of Base Count"
+            plt.figure(figsize=(15, 10))
+            plt.hist(count, min(70, len(count)), color="#135c1c", log=True)
+            plt.xlabel("Base Count", fontsize=14)
+            plt.ylabel("Frequency", fontsize=14)
+            plt.xticks(rotation=90)
+            plt.title(title, fontsize=18)
+            plt.savefig(f"{savedir}{title}.jpeg")
```
Comment (saketkc) [inline pysradb/search.py:570]: Besides saving, can you return the plotting object to the user, for example:
```python
fig, ax = plt.subplots()
ax.scatter(x,y)
return ax
```

Additionally we could also allow for the user to provide us an axes: `ax` that we can then use to create the same scatter plot. This gives the user a lot of flexibility to change titles, labels of axes and play around with colors

File: pysradb/search.py [hunk 4]
```diff
@@ -0,0 +1,1619 @@
...
+                range(len(data.values)),
+                data.values,
+                tick_label=list(data.index),
+                color="#135c1c",
+            )
+            plt.xticks(rotation=90)
+            plt.title(title, fontsize=18)
+            plt.xlabel(axes[0], fontsize=14)
+            plt.ylabel("Frequency", fontsize=14)
+            plt.savefig(f"{savedir}{title}.jpeg")
```
Comment (saketkc) [inline pysradb/search.py:587]: I think the figures should be saved invector format by default. Either to a pdf or a svg. The user can create a jpeg/png if required.

File: pysradb/search.py [hunk 5]
```diff
@@ -0,0 +1,1619 @@
...
+            im = ax.imshow(piv, cmap="Greens")
+            fig.colorbar(im, ax=ax)
+            ax.set_title(title, fontsize=18)
+            ax.set_xticks(range(len(piv.columns)))
+            ax.set_yticks(range(len(piv.index)))
+            ax.set_xticklabels(piv.columns, rotation=90)
+            ax.set_yticklabels(piv.index)
+            ax.set_ylabel(axes[0], fontsize=14)
+            ax.set_xlabel(axes[1], fontsize=14)
+            ax.get_figure().savefig(f"{savedir}{title}.jpeg")
```
Comment (saketkc) [inline pysradb/search.py:624]: save pdf/svg not jpeg

File: pysradb/sradb.py [hunk 6]
```diff
@@ -1265,11 +1262,166 @@ def _srapath_url_srr(srr):
...
+                    "study_accession",
+                    "experiment_accession",
+                    f"run_{i}_accession",
+                    "srapath_url",
+                ]
+                run_df = df.loc[:, df.columns.isin(expected_columns)]
+                run_df = run_df.rename(columns={f"run_{i}_accession": "run_accession"})
+                run_dfs.append(run_df)
+            if run_count == 1:
+                df["srapath_url"] = df.apply(
```
Comment (saketkc) [inline pysradb/sradb.py:1367]: We don't want to select the best url for the user. I think we should output all the URLS and let the user decide what url to choose.  If the user supplies the `--use_ascp` flag we use the aspera ftp downloads. See this notebook for an example of the possible use cases of pysradb download: https://github.com/saketkc/pysradb/blob/master/notebooks/08.pysradb_ascp_multithreaded.ipynb. The download function might need to be changed accordingly.

**Review by saketkc — CHANGES_REQUESTED** [review]

**Comment by saketkc** [comment]
This is  super awesome @bscrow! Many thanks for your contribution and for the awesome work you have done over GSoC2020!
I believe this will be a huge help to a lot of researchers! 

I have left some comments, most of them are minor. It would be great if they can be addressed. Of all, it is particularly important we output all the URLs rather than selecting the best one ourselves.

Great work!

cc @mvdbeek @amalthomas111

File: pysradb/sradb.py (hunk 6 above)
Comment (bscrow) [reply pysradb/sradb.py:1367]: Currently, download is implemented such that the user simply selects yes or no after seeing the list of all URLs. How should the user select which URL column to use?

Comment (saketkc) [reply pysradb/sradb.py:1367]: `download` accepts a [url_col](https://github.com/saketkc/pysradb/blob/5e54c4e37ab15f8f7ac3bc523238c486558770e2/pysradb/sradb.py#L1272) which is also exposed to the cLI. Maybe just use that?

Comment (saketkc) [reply pysradb/sradb.py:1367]: I think you are changing it to None in your code.

File: pysradb/cli.py (hunk 1 above)
Comment (bscrow) [reply pysradb/cli.py]: Resolved in 7a89efd

File: pysradb/cli.py [hunk 7]
```diff
@@ -115,22 +135,12 @@ def download(
     sradb = get_sra_object(db)
     if not srp:
         text = ""
+        # i = 1
```
[the remark this answers is not in the document]
Comment (bscrow) [reply pysradb/cli.py]: Resolved in 7a89efd

File: pysradb/search.py (hunk 3 above)
Comment (bscrow) [reply pysradb/search.py]: resolved in 44ee73c

File: pysradb/search.py (hunk 4 above)
Comment (bscrow) [reply pysradb/search.py]: Resolved in 5104ca6

File: pysradb/search.py (hunk 5 above)
Comment (bscrow) [reply pysradb/search.py]: Resolved in 5104ca6

File: pysradb/sradb.py (hunk 6 above)
Comment (bscrow) [reply pysradb/sradb.py]: Resolved by commits f1678dc to 0d3af0f

**Comment by amalthomas111** [comment]
When using -g it might be a good idea to have a dynamic naming prefix/suffix for the plots. Could use time stamps. Otherwise, plots would be overwritten.

**Comment by amalthomas111** [comment]
For -G, -Y, -Z options it would be great if you could create a file in GitHub or locally which users can refer to, compiling possible options for each of these tags. In the help (-h) options, you can refer to the link of this file or local path.

**Comment by amalthomas111** [comment]
```
pysradb search  -q "single-cell RNA-seq" -g  -D PUBLICATION_DATE 01-01-2008:01-10-2020
```
This command does not work. Gives the error:
ValueError: `bins` must be positive, when an integer

**Comment by amalthomas111** [comment]
pysradb search -d geo -q "single-cell RNA-seq" -m 10K -o test_1ksc

First status bar was showing me 0/100000 [00:00<?, ?it/s]. I mentioned 10K, not 100K. When I mentioned 100, it is showing 1K, a factor of 10 is more.
For both -m = 100 and 10K, I got a connection error:
 http.client.RemoteDisconnected: Remote end closed connection without response. During handling of the above exception, another exception occurred:
Is it NCBI  issue?

**Comment by bscrow** [comment]
> ```
> pysradb search  -q "single-cell RNA-seq" -g  -D  01-01-2008:01-10-2020
> ```
> 
> This command does not work. Gives the error:
> ValueError: `bins` must be positive, when an integer

I've resolved this bug as well as a related bug when the query returns entries without base count - Since none of the entries contains information about base count, the number of bins for a base count histogram will cause the error. 

> When using -g it might be a good idea to have a dynamic naming prefix/suffix for the plots. Could use time stamps. Otherwise, plots would be overwritten.

Thanks for the suggestion! I've implemented it in the new commit

**Comment by bscrow** [comment]
> pysradb search -d geo -q "single-cell RNA-seq" -m 10K -o test_1ksc
> 
> First status bar was showing me 0/100000 [00:00<?, ?it/s]. I mentioned 10K, not 100K. When I mentioned 100, it is showing 1K, a factor of 10 is more. I think this happens with -d geo, not with sra.
> For both -m = 100 and 10K, I got a connection error:
> http.client.RemoteDisconnected: Remote end closed connection without response. During handling of the above exception, another exception occurred:
> Is it NCBI issue?
> 
> I am getting connection/operation time out for almost all m > 100 for db=geo/sra. Need to look into this!

I've debugged the issue of retrieving 10X entries from SRA. 

As for the connection error, I couldn't replicate the error on my side except by running another pysradb search operation from the same IP address while the above process is running, in which case NCBI terminated my connection for exceeding their API limit. Can I check if this is the case when you tested pysradb search?

**Comment by bscrow** [comment]
> For -G, -Y, -Z options it would be great if you could create a file in GitHub or locally which users can refer to, compiling possible options for each of these tags. In the help (-h) options, you can refer to the link of this file or local path.


I've added a short guide for these tags as well as for queries GEO DataSets. This can be accessed via command line using `pysradb search --geo-info` or by calling `GeoSearch.info()` on python.

 21e1c1f

**Comment by saketkc** [comment]
@bscrow a failing example: https://colab.research.google.com/drive/1hN6m7kJ4Xpflvde3wK12Ubzu_Aq3x3qX?usp=sharing

**Comment by bscrow** [comment]
> @bscrow a failing example: https://colab.research.google.com/drive/1hN6m7kJ4Xpflvde3wK12Ubzu_Aq3x3qX?usp=sharing

I've added a check for no search results in 797da82, which should resolve the error message. In order to generate statistics however, instance.search() must be called first to retrieve search results. 

https://colab.research.google.com/drive/1pCmfj-uUDpnBFCXCZoiBw-k82Pi12Otu?usp=sharing

**Comment by bscrow** [comment]
My updated documentation for pysradb search is in the pull request #51 
a live version can be currently found on https://bscrow.github.io/pysradb/commands/search.html

**Pull request merged**